internal/profiling/buffer.go

```diff
@@ -0,0 +1,270 @@
+		// ^uint32(0) is used to denote -1.
+		atomic.AddUint32(&q.acquired, ^uint32(0))
+		goto reloadq
+	}
+
+	// At this point, we're definitely writing to the right queue. That is, one
+	// of the following is true:
+	//   1. No drainer is in execution.
+	//   2. A drainer is in execution and it is waiting at the acquired ==
+	//      written barrier.
+	//
+	// Let's say two Pushes A and B happen on the same queue. Say A and B are
+	// q.size apart; i.e. they get the same index. That is,
+	//
+	//   index_A = index_B
+	//   acquired_A + q.size = acquired_B
+	//
+	// We say "B has wrapped around A" when this happens. In this case, since A
+	// occurred before B, B's Push should be the final value. However, we
+	// accommodate A being the final value because wrap-arounds are extremely
+	// rare and accounting for them requires an additional counter and a
+	// significant performance penalty. Note that the below approach never leads
+	// to any data corruption.
+	index := acquired & q.mask
+	atomic.StorePointer(&q.arr[index], unsafe.Pointer(&x))
+
+	// Allows any drainWait checks to proceed.
+	atomic.AddUint32(&q.written, 1)
+}
+
+// Dereferences non-nil pointers from arr into result. Range of elements from
+// arr that are copied is [from, to). Assumes that the result slice is already
+// allocated and is large enough to hold all the elements that might be copied.
+// Also assumes mutual exclusion on the array of pointers.
+func dereferenceAppend(result []interface{}, arr []unsafe.Pointer, from, to uint32) []interface{} {
+	for i := from; i < to; i++ {
+		// We have mutual exclusion on arr, there's no need for atomics.
+		x := (*interface{})(arr[i])
+		if x != nil {
+			result = append(result, *x)
+		}
+	}
+	return result
+}
+
+// Allocates and returns an array of things Pushed in to the circular buffer.
+// Push order is not maintained; that is, if B was Pushed after A, drain may
+// return B at a lower index than A in the returned array.
+func (cb *circularBuffer) Drain() []interface{} {
+	cb.mu.Lock()
+	defer cb.mu.Unlock()
+
+	qs := make([]*queue, 0)
+	for i := uint32(0); i < numCircularBufferPairs; i++ {
+		qs = append(qs, cb.qp[i].switchQueues())
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(int(numCircularBufferPairs))
+	for i := uint32(0); i < numCircularBufferPairs; i++ {
+		go func(qi uint32) {
+			qs[qi].drainWait()
+
+			// Even though we have mutual exclusion across drainers, this alone
+			// should be done with atomics because this is shared memory that is also
+			// read by Push operations.
+			atomic.StoreUint32(&qs[qi].drainingPostCheck, 1)
+
+			wg.Done()
+		}(i)
+	}
+	wg.Wait()
+
+	result := make([]interface{}, 0)
+	for i := uint32(0); i < numCircularBufferPairs; i++ {
+		if qs[i].acquired < qs[i].size {
+			result = dereferenceAppend(result, qs[i].arr, 0, qs[i].acquired)
+		} else {
+			result = dereferenceAppend(result, qs[i].arr, 0, qs[i].size)
+		}
+	}
+
+	for i := uint32(0); i < numCircularBufferPairs; i++ {
+		qs[i].acquired = 0
```

Comment:
These need to be atomic, too, I believe.
There's a chance that Push() is accessing those concurrently (if the Push() goroutine is not scheduled after loading q, until this)

Reply:
Done.